[kernel 3e/n] Decouple `CDBWrapper` and its kernel users from `ArgsManager`

dongcarl opened the pull request:
This is part of the `libbitcoinkernel` project: #24303, https://github.com/bitcoin/bitcoin/projects/18

This PR is **_NOT_** dependent on any other PRs.

-----

This PR introduces `::Options` structs and `ApplyArgsManOptions` functions for `CDBWrapper` and its wrappers and descendants in `libbitcoinkernel`. Namely `CBlockTreeDB` and `CCoinsViewDB`.

`libbitcoinkernel` code can now instantiate these classes by filling in an `::Options` struct without referencing `gArgs`, and non-`libbitcoinkernel` code can use `ApplyArgsManOptions` with a local `ArgsManager`.

The `::Options` struct and `ApplyArgsManOptions` duo has been used in many previous `libbitcoinkernel` `ArgsManager` decoupling PRs. See: https://github.com/bitcoin/bitcoin/projects/18

dongcarl committed:
DBWrapper: Add constructor using Options struct

The non-Options constructor is kept (it now calls the Options version of
the constructor) so that we can fix callsites one-by-one without
breaking builds, it will be removed later in this patchset.

dongcarl committed:
DBWrapper: Pass in -forcecompactdb via ::Options

dongcarl committed:
Use CDBWrapper::Options ctor for test callers

Also use scopes for freeing temporary CDBWrappers within tests.

dongcarl committed:
DBWrapper: Remove non-Options constructor

Now that the non-Options constructor is no longer used anywhere, remove
it.

dongcarl committed:
Use CDBWrapper::Options ctor for remaining non-test callers

MarcoFalke commented on src/init.cpp:
```diff
@@ -1439,22 +1439,26 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
             .chainparams = chainparams,
             .adjusted_time_callback = GetAdjustedTime,
+            .block_tree_db_opts = {
+                .db_path = gArgs.GetDataDirNet() / "blocks" / "index",
```
1ed5723bc9c69bf2d7c2a902e44f3fc46b1165a7: It is highly illegal to use `gArgs` in init.cpp in new code, when `args` can be used.

dongcarl replied:
Fixed in ed7ae6a08d 😅

maflcko reviewed (approved):
1ed5723bc9c69bf2d7c2a902e44f3fc46b1165a7 🔶

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

1ed5723bc9c69bf2d7c2a902e44f3fc46b1165a7 🔶
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhBHwv/bXhMtBcq5mbrytbt6HR1RqodiVJY8Lco4+XaZJ9SvabncUowi665bc9Q
pTlEAdBYb/RXxJhbRzZe9PTmTpqqZVbOtHaTM/TFfrJMhCSvpuCTKd28Fxav5MMV
En+4Na6IjeZ/NdRT6u2QIM4XgC3AUv3o1rsU8Uw+MHbrAcY+1sn8HuELNxcSg1dL
nUrACZIFIzsiOmNctDP4q+C9bJtnYywdV2DgT7hOED42c5A0cWxanVMvMNMVWAy+
X4m8fnMYjXL4+XmNiMGfrswHscNtWI6jdZd6uIL+kjunQFuPA08S3LVEl2UQKYav
a7G8TuH0OUXIJK6j9VFb1nPotMIbj+bgmvWObBS2R2pRpFglR45IgPmSYq+EnKyk
9pJUT6sdhrGj5RLEHF7tdF4scgqAqpilAmn95TfynEx20KAIhYse8PyRILzdTp2X
M/stGVjV9UyggbepiVfQ0TJC//LlwrGRNFlK8XLf7BHUDikjT8Hc9S3q/HeHtzDZ
zDqoqjm9
=XJoH
-----END PGP SIGNATURE-----
```


</details>

dongcarl commented on src/dbwrapper.cpp:
```diff
@@ -114,48 +114,48 @@ static leveldb::Options GetOptions(size_t nCacheSize)
     return options;
```
Fixed in b37269bedb

dongcarl commented on src/bench/checkblock.cpp:
```diff
@@ -8,6 +8,7 @@
 #include <chainparams.h>
```
Unfortunately not... I think when we get to header pruning there will be a lot of IWYU additions, but it's too early and make the diffs more noisy.

MarcoFalke replied:
Done in https://github.com/bitcoin/bitcoin/pull/25645

dongcarl committed:
Make CChainState::m_coins_view a simple object

Since each CChainState's CoinsViewDB will now be initialized with the
CChainState ctor, we don't need it to be a unique_ptr any more.

Question for reviewers: Is deferring its destruction in Shutdown
codepaths safe?

dongcarl committed:
CoinsViewDB: Add yet unused ::Options, ApplyArgsManOptions

ryanofsky commented on src/dbwrapper.h:
```diff
@@ -236,6 +237,7 @@ class CDBWrapper
                 .in_memory = fMemory,
                 .wipe_existing = fWipe,
                 .obfuscate_data = obfuscate,
+                .do_compact = gArgs.GetBoolArg("-forcecompactdb", false),
```
In commit "CDBWrapper: Pass in -forcecompactdb via ::Options" (d29f09adfaed0abeff0243509cdd49e4f52ebf9a)

gArgs access doesn't really belong in dbwrapper.h header either

ryanofsky commented on src/index/base.cpp:
```diff
@@ -44,9 +44,15 @@ CBlockLocator GetLocator(interfaces::Chain& chain, const uint256& block_hash)
+          .in_memory = f_memory,
+          .wipe_existing = f_wipe,
+          .obfuscate_data = f_obfuscate,
+          .do_compact = gArgs.GetBoolArg("-forcecompactdb", false),
```
In commit "Use CDBWrapper::Options ctor for non-test callers" (c5566d4899d75d9030901092ecf82c97e277d6c2)

Would be good not to add news gArgs calls here and below

ryanofsky commented on src/bitcoin-chainstate.cpp:
```diff
@@ -80,6 +80,10 @@ int main(int argc, char* argv[])
         .chainparams = chainparams,
         .adjusted_time_callback = static_cast<int64_t (*)()>(GetTime),
+        .block_tree_db_opts = {
+            .db_path = gArgs.GetDataDirNet() / "blocks" / "index",
```
In commit "CBlockTreeDB: Add ::Options, init with BlockManager" (961c86a113875366154175677ca4acc326c49e79)

Should probably use `abs_datadir` variable, not `gArgs`

ryanofsky commented on src/txdb.cpp:
```diff
@@ -70,10 +71,17 @@ struct CoinEntry {
+          .in_memory = fMemory,
+          .wipe_existing = fWipe,
+          .obfuscate_data = true,
+          .do_compact = gArgs.GetBoolArg("-forcecompactdb", false),
```
In commit "Use CDBWrapper::Options ctor for non-test callers" (c5566d4899d75d9030901092ecf82c97e277d6c2)

Moving `gArgs.GetBoolArg("-forcecompactdb", false)` call out of cdbwrapper code into txdb code seems like a step backwards or maybe sideways, since txdb code shouldn't be using ArgsManager either

ryanofsky reviewed (approved):
Code review ACK 961c86a113875366154175677ca4acc326c49e79, but I think this would be better if the PR stopped adding ArgsManager::GetArg calls where they don't belong. I don't think there's a good reason to add new calls that will need to be deleted later from validation code when we can just delete them now. I know you aren't really concerned with indexing code, but I don't think it is great to add gArgs usages to indexing code either.

The main culprit here is `gArgs.GetBoolArg("-forcecompactdb", false)` calls and these can be eliminated by adding a `node::ReadDatabaseArgs` function analogous to the `wallet::ReadDatabaseArgs` function. I implemented this in f103f4df57627cd0634b38c99e0aa7da33aedc37 ([tag](https://github.com/ryanofsky/bitcoin/commits/review.25623.1-edit.3)) and also removed gArgs accesses from src/index/ code in the process

dongcarl commented:
@ryanofsky So I think we only add (net) 2 calls to gArgs, one in `src/index/` and one in `src/txdb`. I plan to remove the one in `src/txdb` in a future PR, and add an `ApplyArgsManOptions` function to this PR so that it's easier to remove the reference in `src/index/` in the future.

It is very difficult to maintain behavior and build-ability between commits without having to add oddly placed references to `gArgs` here and there (e.g. the one you pointed out in `dbwrapper.h`), since in order to correctly avoid the oddly placed `gArgs` references we'd have to bubble up the `CDBWrapper::Options` struct all the way to the top in a single commit. This would entail bubbling up both:

1. `CDBWrapper` through `CBlockTreeDB`, through `BlockManager`, through `ChainstateManager`, to init (currently done in this PR)
2. `CDBWrapper` through `CCoinsViewDB`, through `CoinsViews`, through `InitCoinsDB`, through `LoadChainstate`, to init (to be done in a future PR)

in a single commit, which would likely be very hard to review. So we have to have some resting points in these bubbling up chains just so that it's not a huge commit.

I do admit that I can do a better job of adding code comments so that it's clear these `gArgs` references are not meant to stay. I'd also welcome discussing alternative strategies that I might not be thinking of right now.

dongcarl committed:
BlockTreeDB: Construct with {Chain,Block}Manager ctor

Add a CBlockTreeDB::Options member to ChainstateManager::Options, and
pass it down from ChainstateManager's ctor to BlockManager's ctor to
CBlockTreeDB's new ctor that takes only a CBlockTreeDB::Options.

This removes the need to reach into ChainstateManager -> BlockManager ->
CBlockTreeDB right after you initialize the ChainstateManager, which was
ugly. See node/chainstate.cpp and test/util/setup_common.cpp diffs.

Notes:

1. In init.cpp, we needed to move the ChainstateManager initialization
   into the catch_exceptions block, otherwise exceptions thrown by
   the CBlockTreeDB ctor previously caught because they were in
   LoadChainstate would no longer be caught.
2. We also needed to add a data dir member to ChainstateManager::Options
   so that we can determine the default db_path for CBlockTreeDB if none
   is specified (See validation.h hunk). It does seem to make sense that
   ChainstateManager would know about what its data dir is anyway.

dongcarl committed:
BlockTreeDB: Add yet unused ::Options, ApplyArgsManOptions

dongcarl committed:
CoinsViewDB: Construct with ChainState{,Manager} ctor

Add a CCoinsViewDB::Options member to ChainstateManager::Options, store
it as a member in ChainstateManager's ctor, and pass it down to
CChainState's ctor when CSM::{InitializeChainstate,ActivateSnapshot} is
called. CChainState's ctor will then pass it down to CCoinsViews' new
ctor that only takes a CBlockTreeDB::Options.

This removes the need to call InitCoinsDB right after you construct your
CChainState to have a working CChainState at all. In fact,
CChainState::InitCoinsDB is entirely removed in this commit.

dongcarl commented:
Push 961c86a113875366154175677ca4acc326c49e79 -> ccdfad0aadc7c6424f9d1fff1ae7242c8eecb542
- Add commits that also decouples `CCoinsViewDB` from `ArgsManager`
- Reorganize commits

-----

I've thought about it and perhaps it's best to do both `CBlockTreeDB` and `CCoinsViewDB` at once. This reduces the net addition of `gArgs` references to just the one instance in `BaseIndex::DB::DB(...)`, see Ryan's review above.

This is not merge-able yet since `ActivateSnapshot`'s call to `std::unique_ptr<CChainState>` does not take into account `ArgsManager` overrides. I will address this in a future push by adding a `CCoinsViewDB::Options m_coinsdb_opts` member to `ChainstateManager`.

ryanofsky commented on src/kernel/dbwrapper_options.h:
```diff
@@ -0,0 +1,25 @@
+
+namespace kernel {
+
+struct DBWrapperOpts {
+    fs::path db_path;
+    size_t cache_size;
+    bool in_memory = false;
+    bool wipe_existing = false;
+    bool obfuscate_data = false;
+    bool do_compact = false;
+};
```
You can see the other change I suggested here https://github.com/bitcoin/bitcoin/pull/25623#pullrequestreview-1045680511, but I think it simplifies and clarifies code to the do_compact option to a separate struct, as it's a low-level performance tweak that is passed directly from the user to the low-level DB code, unlike the other parameters which are under control of the high level application and indexing code.

It simplifies code by avoiding the need to have 3 near-duplicate structures DBWrapperOpts BlockTreeDBOpts CoinsViewDBOpts that require boilerplate copying individual fields between them. Instead you can 2 structure, with no duplicate fields, that just pass data directly where it needs to go.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/25623#issuecomment-1191907467

> in order to correctly avoid the oddly placed `gArgs` references we'd have to bubble up the `CDBWrapper::Options` struct all the way to the top in a single commit [...] which would likely be very hard to review

I think we will never resolve this disagreement, since we keep on having it! But the two of us have pretty different ideas of what makes things hard to review. A commit like this that makes a focused mechanical change seems extremely easy to review. Things that make a PR hard to review are:

- Commits that mix different types of changes like refactoring & behavior changes, or style cleanups with refactoring
- Commits that are not monotonic improvements, and make some things better, other things worse
- Commits that seem unrelated or don't clearly state motivations
- Multiple commits that churn the same code repeatedly and make it hard to see what direction the PR is moving and actually evaluate whether it is an improvement (not just whether it introduces bugs).

I'd rather look at longer individual commits that are conceptually focused than smaller commits that make less sense individually and make it harder to see a PR's purpose and direction.

This is just some general feedback though. I do think these changes are good and I'm happy to review the code regardless of how changes are split up.

MarcoFalke commented on src/node/blockstorage.h:
```diff
@@ -140,6 +140,8 @@ class BlockManager
     std::unordered_map<std::string, PruneLockInfo> m_prune_locks GUARDED_BY(::cs_main);
 
 public:
+    explicit BlockManager(const CBlockTreeDB::Options& block_tree_db_opts);
```
Wouldn't this be better wrapped into a blockmanopts struct? (Happy to do it myself in another pull if you agree, since I need it elsewhere)

dongcarl replied:
Yeah agree, didn't do it here since this is the only opt, but feel free to do it in your pull!

MarcoFalke replied:
See https://github.com/bitcoin/bitcoin/pull/25781 :black_cat:

dongcarl commented:
re: https://github.com/bitcoin/bitcoin/pull/25623#issuecomment-1201372458

I pretty much agree with everything you said there. I think most times we disagree about whether things are "making things worse" or "a lateral change". Though I will admit I sometimes prioritize the numerical size of commits instead of the "conceptual size", which is something I will watch out for in the future.

Thinking more about my past PRs, I'm surprised you even made it through any of the `g_chainman` deglobalization PRs, perhaps the disdain for mutable globals overweighed the endless lateral move commits I was pushing XD

dongcarl committed:
ChainManOpts: Add coinstip cache size member

Construct ChainstateManager with total_coinstip_cache so that we don't
have to reach into it later in node::LoadChainstate to set it.

Allows us to remove the CacheSizes parameter from node::LoadChainstate
and the member from BasicTestingSetup. Done so in this commit.

dongcarl committed:
ChainManOpts: Add CBlockTreeDB::Options member

Add an ApplyArgsManOptions function that currently just calls the
ApplyArgsManOptions for CBlockTreeDB::Options.

In a future commit in this patchset, this function will also call the
ApplyArgsManOptions for CCoinsViewDB::Options.

dongcarl committed:
Make BlockManager::m_block_tree_db a simple object

Since each BlockManager's m_block_tree_db will now be initialized with
the BlockManager ctor, we don't need it to be a unique_ptr any more.

dongcarl commented:
Push 76381dbc9889472f76c3122535a98e10a6e4b0fc -> 60d3ffa1180666b39efcfecc8ffecadaadcdf844

This push contains many improvements over the one before

- Fixed the `ActivateSnapshot` problem described here: https://github.com/bitcoin/bitcoin/pull/25623#issuecomment-1197374996
- Added a lot of context in commit messages
- Add commit to also make `BlockManager::m_block_tree_db` a simple object member
- Remove `*_in_memory` members from `node::ChainstateLoadOptions` as we subsume them into the respective `::Options` structs
- Initialize `ChainstateManager::m_total_coins{tip,db}_cache` in `ChainstateManager`'s constructor, completely eliminating the `CacheSizes` parameter from `LoadChainstate`
- Merged some commits together so that there are less intermediary codebase states that can confuse reviewers (thanks Ryan!)

dongcarl commented:
Question for reviewers: Is it okay that `m_coins_views` is being destructed later on in the shutdown process after 9292058ae39a130de5de45166fa7e35153fe1ef2?

dongcarl commented:
When you pushed a big change and are waiting to see if the CI passes
![image](https://user-images.githubusercontent.com/3445290/182969261-95b76a5f-1be3-4bfe-8e8c-d66e9b95fc10.png)

MarcoFalke commented:
Could rebase after #https://github.com/bitcoin/bitcoin/pull/25815 ?

dongcarl commented:
Push 60d3ffa118 -> 06bd11d505
- Rebased over #25815
- Updated PR description

ryanofsky reviewed (commented):
Tend to NACK. I think PR is doing some good things, but it is also doing some bad or questionable things. I think the main goal of the PR (decoupling txdb and dbwrapper from ArgsManager) can be achieved more simply. Would suggest starting from #25862 and building from there. Other changes in this PR if they are worth pursuing do not have to be done at the same time as removing ArgsManager calls.

The parts of the PR I think are good are:

- Removing `gArgs` usages from `txdb` and `dbwrapper`.
- Introducing a struct to group `dbwrapper` parameters.

The parts of this PR I think are bad are:

- Making kernel API more complicated. The change makes `bitcoin-chainstate.cpp` example program harder to understand, and inappropriately exposes low level database settings (paths, obfuscation, cache size settings) to kernel applications that will mostly likely break if applications try to set them. I'm in favor of exposing low-level options, but not at cost of making simple applications harder to write. I also think this type of change should be done in a targeted PR, not as an accidental byproduct of internal refactoring.

- Replacing a single `GetBoolArg("-forcecompactdb")` call in dbwrapper code, with 3 separate calls outside of it. This is layer violation exposing low-level database performance options to higher level application code. It also makes it unnecessarily difficult to add new debug/performance options because it now requires adding multiple struct fields and changing multiple layers of code.

- Defining the overlapping fields in 3 different structs (`DBWrapperOpts`, `BlockTreeDBOpts`, `CoinsViewDBOpts`) and requiring boilerplate code to convert between struct types. There isn't a real justification for this complexity right now.

I think #25862 is a clean minimal change that does the work needed to make kernel code not depend on ArgsManager. There may be other changes in this PR that are worth keeping, but I don't think there's a reason to bundle them all together.

dongcarl commented:
Closing in favor of #25862, lets get this done!
